trust-quorum/gfss/src/shamir.rs

```diff
@@ -0,0 +1,255 @@
+use crate::polynomial::{Polynomial, ValidThreshold};
+
+#[derive(Debug, Clone, thiserror::Error, PartialEq, Eq)]
+pub enum SplitError {
+    #[error("splitting requires at least a threshold of 2")]
+    ThresholdToSmall,
+    #[error("total shares {n} must be >= threshold {k}")]
+    TooFewTotalShares { n: u8, k: u8 },
+}
+
+#[derive(Debug, Clone, thiserror::Error, PartialEq, Eq)]
+pub enum CombineError {
+    #[error("must be at least 2 shares to combine")]
+    TooFewShares,
+    #[error("all shares must have distinct x-coordinates")]
+    DuplicateXCoordinates,
+    #[error("all shares must be of equivalent length")]
+    InvalidShareLengths,
+}
+
+/// A set of shares that have been validated
+///
+/// * There are at least 2 shares
+/// * Each share has a unique x-coordinate
+/// * All shares have an equal number of y-coordinates
+pub struct ValidShares<'a>(&'a [Share]);
+
+impl<'a> ValidShares<'a> {
+    pub fn new(shares: &'a [Share]) -> Result<ValidShares<'a>, CombineError> {
+        if shares.len() < 2 {
+            return Err(CombineError::TooFewShares);
+        }
+
+        let len = shares[0].y_coordinates.len();
+        for (i, share) in shares.iter().enumerate() {
+            if share.y_coordinates.len() != len {
+                return Err(CombineError::InvalidShareLengths);
+            }
+            // We only allow constant time comparison of coordinates
+            // and so we must compare each share, with the share that follows it.
+            //
+            // It might be cheaper to store the x-coordinate in a set and check
+            // for duplicates on insert, but we can't do that because we don't
+            // impl `Ord` or `Hash` on `Gf256`.
+            for share2 in shares.iter().skip(i + 1) {
+                if share.x_coordinate.ct_eq(&share2.x_coordinate).into() {
+                    return Err(CombineError::DuplicateXCoordinates);
+                }
+            }
+        }
+
+        Ok(ValidShares(shares))
+    }
+
+    // Return the number of y-coordinates in each share
+    //
+    // This is the length of our secret.
+    pub fn num_y_coordinates(&self) -> usize {
+        self.0[0].y_coordinates.len()
+    }
+}
+
+#[derive(Clone, Zeroize, ZeroizeOnDrop)]
+pub struct Share {
+    pub x_coordinate: Gf256,
+    pub y_coordinates: Vec<Gf256>,
+}
+
+pub struct SecretShares {
+    pub threshold: ValidThreshold,
+    pub shares: Secret<Vec<Share>>,
+}
+
+/// Split a secret into `n` key shares with a threshold of `k` shares.
+pub fn split_secret(
+    secret: &[u8],
+    n: u8,
+    k: u8,
+) -> Result<SecretShares, SplitError> {
+    if k < 2 {
+        return Err(SplitError::ThresholdToSmall);
+    }
+    if n < k {
+        return Err(SplitError::TooFewTotalShares { n, k });
+    }
+
+    // We hardcode this for security. We *may* end up wanting a separate
+    // internal method for property based deterministic tests, but can wait
+    // until we have those.
+    let mut rng = OsRng;
```

Comment:
Time to move all this into a helper method that takes an rng.

Reply:
Resolved by [3e64fac](https://github.com/oxidecomputer/omicron/pull/7891/commits/3e64fac25ca4e65ee7629750331e247b020e16a1)

Thanks @sunshowers